x-pack/plugins/security_solution/docs/adrs/threat_hunting/0001_saving_of_timeline_associated_saved_objects.md

```diff
@@ -0,0 +1,87 @@
+---
+status: proposed
+date: 2024-03-27
+---
+
+# 0001 - Saving of timeline-associated saved objects
+
+## Context and Problem Statement
+
+As described in [#178182](https://github.com/elastic/kibana/issues/178182), the removal of autosave on timeline resulted in a regression in which pinned events and comments on unsaved timelines are lost.
+
+When commenting on an unsaved timeline or an event in a timeline or when pinning an event in a timeline, the pins/comments are lost when the timeline has not been saved before. This used to work in 8.11 and is broken from version 8.12 onwards.
+
+What's causing this bug is that the associated saved objects have a field `timelineId` that connects them to the timeline. When a (pin/note) save request comes in, the server checks for that field and if it doesn't exist, it will create a new timeline on the fly and return its `timelineId` and `timelineVersion` as part of the saved object's response.
+
+https://github.com/elastic/kibana/blob/2df44b9f7f76f3d03f6e32be7f2a39034f97c22e/x-pack/plugins/security_solution/server/lib/timeline/saved_object/notes/saved_object.ts#L138-L148
+
+These two fields are currently not used in the timeline middleware:
+
+https://github.com/elastic/kibana/blob/2df44b9f7f76f3d03f6e32be7f2a39034f97c22e/x-pack/plugins/security_solution/public/timelines/store/middlewares/timeline_note.ts#L60
+
+This gives the false impression that the associated saved object has been stored but it actually has been associated to a different timeline. Subsequent saves of the active timeline (which is not the associated timeline) will then create a new `timelineId`. When the page is reloaded, the associated saved objects will be gone from that timeline.
+
+## Considered Options
+
+- Auto-saving timeline when an a note or pinned event is created
+- Disabling notes and pinned events until the timeline is saved
+- Opening the save modal when clicking the note/pin buttons on an unsaved timeline
+- Caching notes and pinned events locally, saving when timeline is saved
+
+## Decision Outcome
+
+Chosen option: **TBD**
+
+### Confirmation
+
+_PR link will be added later_
+
+## Discussion of options
+
+### Auto-saving timeline when an a note or pinned event is created
+
+In this proposed solution we're bringing back parts of the "auto-save" behaviour. A request to save the associated saved object will create a timeline if the current one hasn't been saved. The request then returns the `timelineId` and `timelineVersion` which will be stored the active timeline.
+This approach has been implemented in https://github.com/elastic/kibana/pull/178212 .
```

Comment:
Now that I have worked on two implementations for this and looked at all options, I'm more convinced that option `Auto-saving timeline when an a note or pinned event is created` is the best approach. It's minimal, has little impact on the UX and the couple extra Untitled timelines are actually not a big deal imo.

Reply:
If it's saved as a draft, I believe users can only have 1 of those at a time, and so the extra timelines would not matter?

Reply:
Yea, if this is the case, this is the best option and product thinks the same

Reply:
Thanks for the hint, I updated the ADR and the PR accordingly :)